### PR TITLE
Open file in new window.

### PR DESCRIPTION
**Pull request by rebornix**
Yay! We love PRs! 🎊

Please include a description of your change & check your PR against this list, thanks:

- [x] Commit message has a short title & issue references
- [x] Commits are squashed 
- [x] It builds and tests pass (e.g `gulp tslint`)

Support `:vsp`, `:vne` and `:vnew`.

This feature is limited to VS Code's group editors, which only support 3 groups at most. Besides, all horizontal splitting commands wont' work in VS Code

### CODE REVIEW

**Comment by johnfn** [comment]
Sorry, would you mind fixing those merge conflicts?

**Comment by rebornix** [comment]
@johnfn I've updated the code to latest :)

**Comment by rebornix** [comment]
BTW, this PR fixes #384 as well.

File: src/cmd_line/commands/file.ts [hunk 1]
```diff
@@ -23,13 +29,48 @@ export class FileCommand extends node.CommandBase {
...
+        if (!active) {
+            return vscode.ViewColumn.One;
+        }
+
+        if (!sideBySide) {
+            return active.viewColumn;
+        }
+
+        switch (active.viewColumn) {
+            case vscode.ViewColumn.One:
```
Comment (johnfn) [inline src/cmd_line/commands/file.ts]: So what is this function doing? What is the sideBySide parameter? This switch appears to get the view column one to the right, but how come it doesn't handle the ViewColumn.Three case? 

It seems to me like you might want two separate functions. getActiveViewColumn() and getViewColumnToRight()

File: src/cmd_line/commands/file.ts [hunk 2]
```diff
@@ -23,13 +29,48 @@ export class FileCommand extends node.CommandBase {
...
+    }
+
     execute(): void {
+        if (!this._arguments.name) {
+            // Open an empty file
+            if (this._arguments.position === FilePosition.CurrentWindow) {
+                vscode.commands.executeCommand("workbench.action.files.newUntitledFile");
+            }
+
+            vscode.commands.executeCommand("workbench.action.splitEditor").then(() => {
```
Comment (johnfn) [inline src/cmd_line/commands/file.ts]: Can you use async/await here?

File: src/cmd_line/subparsers/file.ts [hunk 3]
```diff
@@ -12,6 +12,26 @@ export function parseEditFileCommandArgs(args: string): node.FileCommand {
...
+export function parseEditNewFileInNewWindowCommandArgs(args: string): node.FileCommand {
+    return new node.FileCommand({
+        position: node.FilePosition.NewWindow
+    });
+}
+
+export function parseEditFileInNewWindowCommandArgs(args: string): node.FileCommand {
+    let name = "";
+    if (args) {
+        var scanner = new Scanner(args);
```
Comment (johnfn) [inline src/cmd_line/subparsers/file.ts]: Can you use var, not let?

File: src/cmd_line/commands/file.ts (hunk 1 above)
Comment (rebornix) [inline src/cmd_line/commands/file.ts]: This function is used to get the correct column of group editors when we open new files. If `sideBySide` is true, it means, we want to get the column number of next Group Editor instead of current active editor. But since we can only have 3 columns at most, we need to do some validation on the column number of current editor, if it's already 3, we have to open the file in current editor.

If `sideBySide` is false, what we want to do is get the column of current active editor.

This is what we do in VS Code or any file-related extension (DefaultConfig/UserConfig, Markdown Preview, reStructuredText Preview, etc). Since people might want to tweak it a little bit sometimes, we don't make it an API of any VS Code libraries. 

If we separate it to two functions, we may need move logics to the caller so I'd like to keep it this way.

File: src/cmd_line/commands/file.ts (hunk 2 above)
Comment (rebornix) [inline src/cmd_line/commands/file.ts]: good catch!

File: src/cmd_line/subparsers/file.ts (hunk 3 above)
Comment (rebornix) [inline src/cmd_line/subparsers/file.ts]: Already change it to `var`. That's just my personal preference of using `let` all the time as people always get confused with the scoping of JavaScript's `var`, but here it's totally the same so use `var` is perfect either.

**Comment by rebornix** [comment]
Revise code per comments by @johnfn :)

File: src/cmd_line/commands/file.ts (hunk 1 above)
Comment (johnfn) [inline src/cmd_line/commands/file.ts]: Ok, I understand now. 

> If we separate it to two functions, we may need move logics to the caller so I'd like to keep it this way.

I don't think so. The caller still needs to determine whether to set `sideBySide` to be `true` or `false`. That's the same as determining whether to call getActiveViewColumn() and getViewColumnToRight(). I suggest splitting the function since those feel to me like separate bits of functionality.

File: src/cmd_line/subparsers/file.ts [hunk 4]
```diff
@@ -9,7 +9,7 @@ export function parseEditFileCommandArgs(args: string): node.FileCommand {
     }
 
     var scanner = new Scanner(args);
-    let name = scanner.nextWord();
+    var name = scanner.nextWord();
```
Comment (johnfn) [inline src/cmd_line/subparsers/file.ts]: Wait, why did you change _more_ `let`s to `var`s? That's the opposite of what I want! 😉

File: src/cmd_line/commands/file.ts (hunk 1 above)
Comment (rebornix) [inline src/cmd_line/commands/file.ts]: Sure, will do that.

File: src/cmd_line/subparsers/file.ts (hunk 4 above)
Comment (rebornix) [inline src/cmd_line/subparsers/file.ts]: Now I'm a little bit confused. Previously I was thinking your way of deciding when to use `let` or `var` is

* If it's inside a block (`{ }`, for loop), then we can use `let`
* If it's just in the very top level of the function, then we use `var`

Since JavaScript does hoisting and moves all declaration to the beginning this `var name` here is almost the same as `var name = ""` below in `parseEditFileInNewWindowCommandArgs` (https://github.com/VSCodeVim/Vim/pull/404/files#diff-16bc6b0f3306c3b8bdc5733a584411ffR27). You must have solid reason for this but I'm sorry that I'm not that clear now, do you mind pointing it out?

Comment (johnfn) [inline src/cmd_line/subparsers/file.ts]: My thoughts about variable types:

* Use `const` as much as possible
* When `const` is impossible (the variable gets changed), use `let`

Even though `let` and `var` are equivalent in this case, `let` is generally superior to `var` and there is no disadvantage to using `let; therefore we should always use `let`.

I'll whip up a style guide soon. :smile:

Comment (rebornix) [inline src/cmd_line/subparsers/file.ts]: oh your suggestion is using `let` whenever possible.

so previously when you said `use var, not let` in ` parseEditFileInNewWindowCommandArgs` is kind of typo? :)

Comment (johnfn) [inline src/cmd_line/subparsers/file.ts]: OOPS! Yes, that was a typo! Sorry!

Comment (rebornix) [inline src/cmd_line/subparsers/file.ts]: perfect. We are using `let` totally the same way. `let` always :) I'll give it a quick turnaround.

**Comment by rebornix** [comment]
@johnfn and I misunderstood each other about the usage of `let` and `var` but actually we are saying the same thing: use `let` always possible.

Fixed all existing comments, I suppose it's a good time to merge.

**Comment by johnfn** [comment]
@rebornix 2 things left:

* Split that function into 2
* Switch that one `var` to a `let` that I commented on (sorry!)

Then it should be good to go!

**Comment by rebornix** [comment]
I must be drunk last time I checked in code, I thought I already fixed those issues ...

Follow your `Better Naming` suggestion and replace all `var` with `let` in my PR. Maybe we can create some gulp task `fix-var` similar to `fix-whitespace` :)
